# GlobalPowerPlant_Project is a Regression and Classification Problem

In [1]:
print(('*'*127))
print("                                                      Global Power Plant                                                                 ")
print(('*'*127))

*******************************************************************************************************************************
                                                      Global Power Plant                                                                 
*******************************************************************************************************************************


In [2]:
import pandas as pd
import numpy as np
import seaborn as sn


from matplotlib import pyplot as plt
import warnings
warnings.filterwarnings('ignore')

df=pd.read_csv("https://raw.githubusercontent.com/wri/global-power-plant-database/master/source_databases_csv/database_IND.csv")

pd.set_option('display.max_rows', 950)
pd.set_option('display.max_columns', 28)


df

country country_long                                               name  \
0       IND        India                                   ACME Solar Tower   
1       IND        India                                ADITYA CEMENT WORKS   
2       IND        India                           AES Saurashtra Windfarms   
3       IND        India                                        AGARTALA GT   
4       IND        India                                       AKALTARA TPP   
5       IND        India                                       AKRIMOTA LIG   
6       IND        India                                             ALIYAR   
7       IND        India                                    ALLAIN DUHANGAN   
8       IND        India                                        ALMATTI DAM   
9       IND        India                                        AMAR KANTAK   
10      IND        India                                      AMARAVATI TPP   
11      IND        India                             ANANDPUR SAHIB ST-I&II   
12      IND        India                                        ANAPARA "C"   
13      IND        India                                             ANDHRA   
14      IND        India                                  ANOOPGARH ST I&II   
15      IND        India                                             ANPARA   
16      IND        India                                            ANTA GT   
17      IND        India                                            ANUPPUR   
18      IND        India                                         ANUPUR TPP   
19      IND        India                                  ARVINDNAGAR SUGAR   
20      IND        India                                    ASHOKNAGAR MILL   
21      IND        India                                       ATHANI SUGAR   
22      IND        India                                         AURAIYA GT   
23      IND        India                                AVANTHA BHANDAR TPP   
24      IND        India                               AWARPUR CEMENT PLANT   
25      IND        India                                           Abhijeet   
26      IND        India                          Amarsar Solar Power Plant   
27      IND        India                                  Anabura Wind Farm   
28      IND        India                Ananthapur - Meil Solar Power Plant   
29      IND        India                              Andrha Lake Wind Farm   
30      IND        India                               Anikkadavu Wind Farm   
31      IND        India                             Arasinagundi Wind Farm   
32      IND        India                 Askandra - DREPL Solar Power Plant   
33      IND        India  Askandra - Electromech Maritech Solar Power Plant   
34      IND        India              Askandra - Finehope Solar Power Plant   
35      IND        India                 Askandra - Khaya Solar Power Plant   
36      IND        India                Askandra - Newton Solar Power Plant   
37      IND        India               Askandra - Saidham Solar Power Plant   
38      IND        India                Askandra - Vasavi Solar Power Plant   
39      IND        India                                           BADARPUR   
40      IND        India                                       BAGALKOT ICP   
41      IND        India                                       BAGLIHAR HEP   
42      IND        India                                    BAGLIHAR HEP-II   
43      IND        India                                         BAIRA SIUL   
44      IND        India                                          BAKRESWAR   
45      IND        India                                          BALCO TPP   
46      IND        India                                           BALIMELA   
47      IND        India                               BALLARPUR BILT POWER   
48      IND        India                                     BAMANIA CEMENT   
49      IND        India                              

In [3]:
df.shape

(907, 27)

* Here we that there is 907 Rows and 27 Columns for this Dataset

In [4]:
df['primary_fuel'].value_counts()

Coal       258
Hydro      251
Solar      127
Wind       123
Gas         69
Biomass     50
Oil         20
Nuclear      9
Name: primary_fuel, dtype: int64

* Here we observe for each primary_fuel the value counts is highest for Coal and Lowest for Nuclear and This Column is One of Our Target Variable

In [5]:
df.isnull().sum()

country                       0
country_long                  0
name                          0
gppd_idnr                     0
capacity_mw                   0
latitude                     46
longitude                    46
primary_fuel                  0
other_fuel1                 709
other_fuel2                 906
other_fuel3                 907
commissioning_year          380
owner                       565
source                        0
url                           0
geolocation_source           19
wepp_id                     907
year_of_capacity_data       388
generation_gwh_2013         907
generation_gwh_2014         509
generation_gwh_2015         485
generation_gwh_2016         473
generation_gwh_2017         467
generation_gwh_2018         459
generation_gwh_2019         907
generation_data_source      458
estimated_generation_gwh    907
dtype: int64

* Since we observe Columns other_fuel1, other_fuel2, other_fuel3, wepp_id, generation_gwh_2013, generation_gwh_2019, estimated_generation_gwh have many Missing Values as we Treat them using Mean, Median or Mode

In [6]:
df.dtypes

country                      object
country_long                 object
name                         object
gppd_idnr                    object
capacity_mw                 float64
latitude                    float64
longitude                   float64
primary_fuel                 object
other_fuel1                  object
other_fuel2                  object
other_fuel3                 float64
commissioning_year          float64
owner                        object
source                       object
url                          object
geolocation_source           object
wepp_id                     float64
year_of_capacity_data       float64
generation_gwh_2013         float64
generation_gwh_2014         float64
generation_gwh_2015         float64
generation_gwh_2016         float64
generation_gwh_2017         float64
generation_gwh_2018         float64
generation_gwh_2019         float64
generation_data_source       object
estimated_generation_gwh    float64
dtype: object

* Here we observe a Majority of Float-type variables

# Dataset Descirption

The happiness scores and rankings use data from the Gallup World Poll.

The scores are based on answers to the main life evaluation question asked in the poll. This question, known as the Cantril ladder, asks respondents to think of a ladder with the best possible life for them being a 10 and the worst possible life being a 0 and to rate their own current lives on that scale. 

The scores are from nationally representative samples for the years 2013-2016 and use the Gallup weights to make the estimates representative. 

A hypothetical country that has values equal to the world’s lowest national averages for each of the six factors. 

They have no impact on the total score reported for each country, but they do explain why some countries rank higher than others.

# Attributes

The reports review the state of happiness in the world today

---Country	
---Region	
---Happiness Rank	
---Happiness Score	
---Standard Error	
---Economy (GDP per Capita)	
---Family	
---Health (Life Expectancy)	
---Freedom	
---Trust (Government Corruption)	
---Generosity	
---Dystopia Residual# 

# Data Cleaning

# To Treat Missing Values

In [7]:
df['latitude']=df['latitude'].fillna(np.mean(df['latitude']))
df['longitude']=df['longitude'].fillna(np.mean(df['longitude']))
df['commissioning_year']=df['commissioning_year'].fillna(np.mean(df['commissioning_year']))
df['year_of_capacity_data']=df['year_of_capacity_data'].fillna(np.mean(df['year_of_capacity_data']))
df['generation_gwh_2014']=df['generation_gwh_2014'].fillna(np.mean(df['generation_gwh_2014']))
df['generation_gwh_2015']=df['generation_gwh_2015'].fillna(np.mean(df['generation_gwh_2015']))
df['generation_gwh_2016']=df['generation_gwh_2016'].fillna(np.mean(df['generation_gwh_2016']))
df['generation_gwh_2017']=df['generation_gwh_2017'].fillna(np.mean(df['generation_gwh_2017']))
df['generation_gwh_2018']=df['generation_gwh_2018'].fillna(np.mean(df['generation_gwh_2018']))


In [8]:
# We drop those columns which has absolute No data

df.drop(columns=['other_fuel1', 'other_fuel2', 'other_fuel3', 'wepp_id', 'estimated_generation_gwh', 'generation_gwh_2019', 'generation_gwh_2013'],axis=1,inplace=True) 

In [9]:
# To Encode those Columns for later Treating Missing Values

from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()

df["owner"] = lb.fit_transform(df["owner"])
df["geolocation_source"] = lb.fit_transform(df["geolocation_source"])
df["generation_data_source"] = lb.fit_transform(df["generation_data_source"])

# To Perform Ordinal Encoding

In [10]:
from sklearn.preprocessing import OrdinalEncoder

oe=OrdinalEncoder()

for i in df.columns:
    if df[i].dtypes=="object":
        df[i]=oe.fit_transform(df[i].values.reshape(-1,1))

In [ ]:
df

In [ ]:
df.isnull().sum()

# Checking VIF ( Variance Inflation Factor )

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

def calculate_vif(df):
    vif=pd.DataFrame()
    vif["F"] = df.columns
    vif["VIF_Value"] = [variance_inflation_factor(df.values, i)for i in range(df.shape[1])]
    return(vif)

F=df.iloc[:,0:4]
F.head()

In [ ]:
G=df.iloc[:,8:]
G.head()

In [ ]:
calculate_vif(F)

In [ ]:
calculate_vif(G)

* In the above VIF score we see that Contry and Country Long has NaN values since which we could Remove any one Column

In [ ]:
df.drop(columns=['country'],axis=1,inplace=True) 

# Outlier Check and Removal

In [ ]:
df1 = df[['country_long', 'name', 'gppd_idnr', 'capacity_mw', 'latitude', 'longitude', 'primary_fuel', 'commissioning_year', 'owner', 'source', 'url']]

In [ ]:
Columns = df1.columns.values
ncolumns = 14
nrows = 10
plt.figure(figsize=[40,15])
for i in range(0, len(Columns)):
    plt.subplot(nrows,ncolumns,i+1)
    sn.boxplot(df1[Columns[i]], color='green', orient='H', palette='husl')
    plt.tight_layout()

In [ ]:
df2 = df[['geolocation_source','year_of_capacity_data','generation_gwh_2014','generation_gwh_2015','generation_gwh_2016','generation_gwh_2017','generation_gwh_2018','generation_data_source']]

In [ ]:
Columns = df2.columns.values
ncolumns = 14
nrows = 10
plt.figure(figsize=[40,15])
for i in range(0, len(Columns)):
    plt.subplot(nrows,ncolumns,i+1)
    sn.boxplot(df2[Columns[i]], color='green', orient='H', palette='husl')
    plt.tight_layout()